zerver/lib/message.py

```diff
@@ -1282,6 +1291,94 @@ def remove_message_id_from_unread_mgs(state: RawUnreadMessagesResult, message_id
     state["mentions"].discard(message_id)
 
 
+def format_unread_message_details(
+    user_id: int,
+    raw_unread_data: RawUnreadMessagesResult,
+) -> Dict[str, Dict[str, Any]]:
+    unread_data = {}
+
+    for message_id, private_message_details in raw_unread_data["pm_dict"].items():
+        other_user_id = private_message_details["other_user_id"]
+        if other_user_id == user_id:
+            user_ids = []
+        else:
+            user_ids = [other_user_id]
+
+        message_details: Dict[str, object] = {
+            "type": "private",
+            "user_ids": user_ids,
+        }
+        if message_id in raw_unread_data["mentions"]:
+            message_details["mentioned"] = True
+        unread_data[str(message_id)] = message_details
+
+    for message_id, stream_message_details in raw_unread_data["stream_dict"].items():
+        if message_id in raw_unread_data["unmuted_stream_msgs"]:
+            unmuted_stream_msg = True
+        else:
+            unmuted_stream_msg = False
+
+        message_details = {
+            "type": "stream",
+            "stream_id": stream_message_details["stream_id"],
+            "topic": stream_message_details["topic"],
+            # Clients don't need this detail, but we need it internally for apply_events.
+            "unmuted_stream_msg": unmuted_stream_msg,
+        }
+        if message_id in raw_unread_data["mentions"]:
+            message_details["mentioned"] = True
+        unread_data[str(message_id)] = message_details
+
+    for message_id, huddle_message_details in raw_unread_data["huddle_dict"].items():
+        user_ids = [int(s) for s in huddle_message_details["user_ids_string"].split(",")]
+        user_ids = [uid for uid in user_ids if uid != user_id]
+        user_ids.sort()
```

Comment:
Maybe worth a brief block comment about the design of this format -- a sorted list of user IDs not containing the current user, just to make it easy to verify that this implementation is as intended.